man/lpc.1
.ds VE LPRng-3.1.2
.TH LPC 1 \*(VE "LPRng"
.ig
$Id: lpc.1,v 3.4 1997/01/31 22:42:36 papowell Exp $
Mon Jul 17 20:14:17 PDT 1995 Patrick Powell
..
.SH NAME
lpc \- line printer control program
.SH SYNOPSIS
.B lpc
[
.B \-A
]
[
.BI \-D debugopts
]
[
.BI \-P printer
]
[
.B \-V
]
[ command [ argument ... ] ]
.SH DESCRIPTION
.I Lpc
is used by the system administrator to control the
operation of the line printer system.  
For each line printer configured in 
.I /etc/printcap 
or the printcaps specified in the configuration file 
.I lpd.conf
(see lpd.conf(5)),
.I lpc
may be used to:
.IP \(bu 3
disable or enable a single or all printer,
.IP \(bu 3
disable or enable a single printer's or all printer's spooling queue,
.IP \(bu 3
rearrange the order of jobs in a spooling queue (moving jobs to the
top or to the end of a queue or changing their priority in the queue),
.IP \(bu 3
find the status of printers, and their associated
spooling queues and printer dameons,
.IP \(bu 3
start and stop printer servers for a queue with multiple print servers,
.IP \(bu 3
hold and release a specific job in a printer queue,
.IP \(bu 3
redirect printing to another printer
.IP \(bu 3
restart a printer job after having solved printer problems and let it
print from the beginning, 
.PP
Without any arguments,
.I lpc
will prompt for commands from the standard input.
If arguments are supplied,
.IR lpc
interprets the first argument as a command and the remaining
arguments as parameters to the command.
Permission to use spool queue control commands is determined by the
printer permissions file
(See FILES).
.SH OPERATION
.PP
The operation of each spool queue is controlled by a
.I "spoolcontrol"
file which has a set of keyword options and values.
(Other print spoolers have used permissions bits of directories and files
for similar purposes).
In addition,
each job in the spool queue can have a
.I "holdfile"
which contains detailed information on how the spooler is to treat the
particular job.
The server will update these files with status and other information as the
job is processed.
The
.I lpc
command operates by sending requests to the
.I lpd
server process to update the information in these files and to signal
server processes that the information has been updated.
.SH OPTIONS
.IP "\fB\-A\fR" 5
Use authentication when sending command to server.
.IP "\fB\-P\fIprinter\fR" 5
Printer spool queue to operate on.
if no name is given,
the -P option,
the printer selected by the value of the PRINTER enviroment variable,
or the first entry in the printcap file will be used.
When all printers are selected,
the LPC command will first attempt to find the
printcap entry for printer ``all'';
the
.I all
printcap field value will be a list of printers to control. 
If there is no printcap entry,
then the request is directly forwarded to the default LPD host,
which will do a lookup for all of the available printers.
By specifying ``all@host'',
the user can control all printers on a given host.
.IP "\fB\-V\fR" 5
Print program version information
.PP
.IP "\fB\-D\fIdebugopts\fP" 5
Debugging is controlled using the
.B \-D
option. This accepts a comma-separated list of debugging
settings. These settings take one of two forms: 
.B facility=value , 
or
.B value 
to set an overall default value.
The form -D= will display a list of debugging options.
.SH COMMANDS
.PP
The following is the alphabetical list of recognized commands.
The jobcontrol or holdfile fields effective are listed as
part of the command.
.TP
?
.br
Help.
Also, any unrecognized input produces a help listing.
.TP
abort { all | printer* }
.br
Kill the active job and disables unspooling.
The active job will not be deleted from the queue.
.br
Action: kills server process, updates spool control file
.I printing_disabled
field.
.TP
holdall { all | printer* }
.br
Turn on automatic job holding.
Jobs will be held until released with the hold command,
or holdall turned off with the noholdall command.
The :ah: printcap flag can enable automatic holding as well.
.br
Action: updates spool control file
.I holdall
field.
.TP
debug j all | printer } [string | off ]
.br
Set the debugging string for the specified printer.
This is a diagnostic aid and should be used with caution;
refer to the reference manual for details of the debugging string format.
In general,
you can use:
.br
.I INTEGER
\- the integer number sets the general debugging level
.br
keyword=value
.br
\- sets a debugging variable to a specified value
.br
keyword
\- sets a debugging flag
.br
Action: updates spool control file
.I debug
field.
.TP
disable { all | printer }
.br
Disable spooling to the specified spool queues.
See
.I enable.
.br
Action: updates spool control file
.I printing_disabled
field.
.TP
class {all | printer } (off | classlist | X=globmatch)
.br
Controls the class of jobs currently being printed.
The
.I off
option removes any class restrictions.
The classlist option is a list of classes;
for example A,B,C would release classes A, B, and C for printing.
The X=globmatch form performs a match against a control file entry
starting with the indicated capital letter.
For example,  J=*form1* would allow jobs which had the string
form1 in their job title to be printed.
.TP
down {all | printer }
.br
Disable both queuing and printing for all printers or the selected printer.
See
.I up.
.br
Action: updates spool control file
.I printing_disabled
and
.I spooling_disabled
fields.
.TP
enable { all | printer }
.br
Enable spooling for all printers or the selected printer.
.br
Action: updates spool control file
.I spooling_disabled
field.
.TP
exit or quit
.br
terminate LPC program
.TP
help
.br
Print a short help message.
.TP
hold printer { jobid }
.br
Holds the specified printer job (or jobs) in the queue.
The job will not be
printed even if spooling and printing is enabled. It must be released
with the relase command.
The holdall command
or :ah: printcap flag can enable automatic holding of spooled jobs.
.br
Action: updates job holdfile
.I hold
field with time that job was held.
A 0 value release job for printing.
.TP
kill { all | printer ... }
.br
Do an abort command followed by a start command.
This is a quick way to kill off a server that has problems.
Note that due to race conditions and delays,
that the start command may not be effective this action should
be monitored to ensure that the job restarts.
.TP
lpd
.br
determines the LPD daemon process is running,
This is handy to determine if the LPD daemon was killed or
aborted due to abnormal conditions.
.TP
lpq printer [options]
.br
Run
.I lpq
from inside the 
.I lpc
program.
.TP
lprm printer [options]
.br
Run
.I lprm
from inside the 
.I lpc
program.
.TP
move printer { jobid }  destinationPrinter
.br
Send the specified jobs to the destination printer and remove them
from the printer queue.
.br
Action: updates the job holdfile
.I move
field with the destination and starts a server process to do the
job transfer.
.TP
noholdall { all | printer* }
.br
Turn off automatic job holding.
See holdall command.
.TP
quit or exit
.br
terminate LPC program
.TP
redirect [printer [destinationPrinter | off ]]
.br
redirect the jobs in the printer queue to another printer
or turn redirection off.
.br
Action: updates spool control file
.I redirect
field.
.TP
release [printer [jobid]]
.br
Releases the selected job for printing.
.TP
reread [ printer ]
.br
Sends a request to the LPD server for the printer to reread the
configuration and printcap information.
This is equivalent to using
kill -HUP serverpid,
but can be done for servers on remote hosts.
.TP
start { all | printer }
.br
Start the printer.
This is useful when some abnormal condition causes the server to
terminate unexpectedly leaving jobs in the queue.
.I Lpq
will report that there is no daemon present when this condition occurs. 
.TP
status { all | printer }
Display the status of daemons and queues on the local machine.
.TP
stop { all | printer }
.br
Disable any further unspooling after the current job completes.
.TP
topq printer [ jobid ]
.br
Place the selected jobs at the top of the printer queue.
.TP
up {all | printer ...}
Enables queuing and printing for the specified or all printers.
Privileged.
.SH JOBIDS
.PP
The LPQ command displays a job identifier for each job,
which can be used in LPC commands to identify a specific job.
In the commands descriptions above,
the jobid can be a user name,
a job number,
a job identifier,
or a glob pattern.
The glob pattern is matched against the job identifier.
In command which have an optional jobid,
if none is specified then the first printable job in the queue is acted on.
.SH FILES
.PP
The files used by LPRng are set by values in the
printer configuration file.
The following are a commonly used set of default values.
.nf
.ta \w'/var/spool/lpd/printcap.<hostname>           'u
/etc/lpd.conf		LPRng configuration file
/etc/printcap		printer description file
/etc/lpd.perms	printer permissions
/var/spool/printer*		spool directories
/var/spool/printer*/printer	lock file for queue control
/var/spool/printer*/control.printer	queue control
/var/spool/printer*/active.printer	active job
/var/spool/printer*/log.printer	log file
.fi
.SH "SEE ALSO"
lpd.conf(5),
lpc(8),
lpd(8),
lpr(1),
lpq(1),
lprm(1),
printcap(5),
lpd.perms(5),
pr(1).
.SH DIAGNOSTICS
.nf
Most of the diagnostics are self explanatory.
If you are puzzled over the exact cause of failure,
set the debugging level on (-D5) and run again.
The debugging information will 
help you to pinpoint the exact cause of failure.
.fi
.SH "HISTORY"
.LP
LPRng is a enhanced printer spooler system
with functionality similar to the Berkeley LPR software.
In 1988 Patrick Powell released
the PLP (Public Line Printer) software,
which went through several evolutions.
Justin Mason (jmason@iona.ie)
generated PLP4.0 from several older releases of PLP.
In 1992 Patrick Powell
release LPRng,
a completely redesigned and newly written version of the software.
.LP
The LPRng mailing list is plp@iona.ie;
subscribe by sending mail to plp-request@iona.ie with
the word subscribe in the body.
The software is available from ftp://iona.ie/pub/LPRng.
.LP
LPRng is distributed under the GNU software license for non-commercial
use,
the Artistic License for limited commercial use. 
Commerical support and licensing is available through
Patrick Powell <papowell@sdsu.edu>.


NEWS
---------------------------------------------------------------------
Version LPRng-3.8.A - not yet released
---------------------------------------------------------------------

This is the first release not done by Patrick Powell but
by the new team to maintain LPRng. The project is hosted
on sourceforge, take a look at
http://sourceforge.net/projects/lprng/
and join us on the new mailing-list at
http://lists.sourceforge.net/mailman/listinfo/lprng-devel

The biggest change is the switch to using automake to generate
the Makefile.im and cleaning the autoconf rules a bit. Most
things should work just as before, with the following exceptions:

- make install no longer tries to remove previous installed
  printing systems and to modify your system to run LPRng, you
  will have to do this yourself. Example files with descriptions
  can be found in the conf/ subdirectory of the source.
- binaries are currently not made suid even if requested,
  that might be reimplemented in the future.
- when configuring with --enable-nls less time is invested in
  finding a gettext implementation. You might need to give it
  some more flags when using a gettext not in your libc.

There is a new configure option --disable-remote to change the default
to not listen via tcp and not accepting requests from remote hosts.

Otherwise a lot of cleanup has happened. Many compiler warnings
quieted, prototypes added, format strings synced with format
of arguments, some code and manpages with strange license situation
replaced, local functions declared static, moving some code around
to not have to link all lpd code in all other binaries...

---------------------------------------------------------------------
Changes in versions until LPRng-3.8.28 can
be found in the file CHANGES.
If that file did not come with your LPRng, you should be able to download
it from: http://lprng.cvs.sourceforge.net/*checkout*/lprng/lprng/CHANGES


src/LPC/lpc.c
/***************************************************************************
 * LPRng - An Extended Print Spooler System
 *
 * Copyright 1988-1995 Patrick Powell, San Diego State University
 *     papowell@sdsu.edu
 * See LICENSE for conditions of use.
 *
 ***************************************************************************
 * MODULE: lpc.c
 * PURPOSE:
 **************************************************************************/

/***************************************************************************
 * SYNOPSIS
 *      lpc [ -PPrinter ] [command]
 * commands:
 *
 *   status [printer]     - show printer status (default is all printers)
 *   start [printer]      - start printing
 *   stop [printer]       - stop printing
 *   enable [printer]     - enable spooling
 *   disable [printer]    - disable spooling
 *   abort [printer]      - stop printing, kill server
 *   kill [printer]       - stop printing, kill server, restart printer
 *
 *   topq printer (user [@host] | host | jobnumer)*
 *   hold printer (all | user [@host] | host |  jobnumer)*
 *   release printer (all | user [@host] | host | jobnumer)*
 *
 *   lprm printer [ user [@host]  | host | jobnumber ] *
 *   lpq printer [ user [@host]  | host | jobnumber ] *
 *   lpd                  - PID of LPD server
 *
 * DESCRIPTION
 *   lpc sends a  request to lpd(8)
 *   and reports the status of the command
 ****************************************************************************
 *
Implementation Notes
Patrick Powell Wed Jun 28 21:28:40 PDT 1995

The LPC program is an extremely simplified front end to the
LPC functionality in the server.  The commands send to the LPD
server have the following format:

\6printer user command options

If no printer is specified, the printer is the default from the
environment variable, etc.

*/

#include "lpc.h"
#include "lp_config.h"
#include "control.h"
#include "setuid.h"
#include "getprinter.h"

static char *const _id =
"$Id: lpc.c,v 3.2 1996/09/09 14:24:41 papowell Exp papowell $";

/***************************************************************************
 * main()
 * - top level of LPP Lite.
 *
 ****************************************************************************/


void Send_lpcrequest(
	char *user,					/* user name */
	char **options,				/* options to send */
	int connect_timeout,		/* timeout on connection */
	int transfer_timeout,		/* timeout on transfer */
	int output,					/* output file descriptor */
	int action					/* action */
	);

int main(int argc, char *argv[], char *envp[])
{
	char *s;
	char msg[ LINEBUFFER ];
	int action;

	/*
	 * set up the user state
	 */
	Interactive = 1;
	Initialize();


	/* set signal handlers */
	(void) plp_signal (SIGPIPE, (plp_sigfunc_t)SIG_IGN);
	(void) plp_signal (SIGHUP, cleanup);
	(void) plp_signal (SIGINT, cleanup);
	(void) plp_signal (SIGQUIT, cleanup);
	(void) plp_signal (SIGTERM, cleanup);


	/* scan the argument list for a 'Debug' value */

	Opterr = 0;
	Get_debug_parm( argc, argv, LPC_optstr, debug_vars );

	/* Get configuration file information */
	Parsebuffer( "default configuration", Default_configuration,
		lpc_config, &Config_buffers );

	/* get the configuration file information if there is any */
    if( Allow_getenv ){
		if( UID_root ){
			fprintf( stderr,
			"%s: WARNING- LPD_CONF environment variable option enabled\n"
			"  and running as root!  You have an exposed security breach!\n"
			"  Recompile without -DGETENV or do not run clients as ROOT\n",
			Name );
		}
		if( (s = getenv( "LPD_CONF" )) ){
			Client_config_file = s;
		}
    }

	DEBUG0("main: Configuration file '%s'",
		Client_config_file?Client_config_file:"NULL" );

	Getconfig( Client_config_file, lpc_config, &Config_buffers );

	if( Debug > 5 ) dump_config_list( "LPC Configuration", lpc_config );


	/* get the fully qualified domain name of host and the
		short host name as well
		FQDN - fully qualified domain name
		Host - actual one to use in H fields
		ShortHost - short host name
		NOTE: on PCs this will be the IP address
	*/

	Get_local_host();

	/* expand the information in the configuration file */
	Expandconfig( lpc_config, &Config_buffers );

	if( Debug > 4 ) dump_config_list( "LPC Configuration After Expansion",
		lpc_config );

	Logname = Get_user_information();

	/* scan the input arguments, setting up values */
	Opterr = 1;
	Get_parms(argc, argv);      /* scan input args */

	/* get the printer name */
	Get_printer();
    if( (RemoteHost == 0 || *RemoteHost == 0) ){
		RemoteHost = 0;
        if( Default_remote_host && *Default_remote_host ){
            RemoteHost = Default_remote_host;
        } else if( FQDNHost && *FQDNHost ){
            RemoteHost = FQDNHost;
        }
    }
	if( RemoteHost == 0 ){
		Diemsg( "No remote host specified" );
	}
    if( (RemotePrinter == 0 || *RemotePrinter == 0) ){
		RemotePrinter = Printer;
	}

	DEBUG4("lpc: printer '%s', remote host '%s'", Printer, RemoteHost );
	DEBUG4("lpc: Optind '%d', argc '%d'", Optind, argc );
	if( Optind < argc ){
		/* we have a command line */
		Send_lpcrequest( Logname, &argv[Optind],
			Connect_timeout, Send_timeout, 1, 0 );
	} else while(1){
		int tokencount, i;
#define MAXTOKEN 20
		struct token tokens[ MAXTOKEN ];
		char *lineargs[ MAXTOKEN+1 ];
	
		fprintf( stdout, "lpc>" );
		fflush( stdout );
		if( fgets( msg, sizeof(msg), stdin ) == 0 ) break;
		if( (s = strchr( msg, '\n' )) ) *s = 0;
		DEBUG4("lpc: '%s'", msg );
		tokencount = Crackline( msg, tokens, MAXTOKEN );
		if( tokencount == 0 ) continue;
		if( tokencount >= MAXTOKEN ){
			fprintf( stdout, "Too many parameters\n" );
			continue;
		}
		for( i = 0; i < tokencount;  ++i ){
			lineargs[i] = tokens[i].start;
			tokens[i].start[tokens[i].length] = 0;
		}
		lineargs[i] = 0;
		s = lineargs[0];
		if( strcasecmp(s,"exit") == 0 || s[0] == 'q' || s[0] == 'Q' ){
			break;
		}
		action = Get_controlword( s );
		if( action == 0 ){
			use_msg();
			continue;
		}
		Send_lpcrequest( Logname, lineargs,
			Connect_timeout, Send_timeout, 1, action );
	}
	exit(0);
}
